File: R/exponpen.R
exponpen <- function(basisfd, Lfd=2)
{

#  Computes the Exponential penalty matrix.
#  Argument:
#  BASISFD ... a basis.fd object of type "expon"
#  LFD     ... either the order of derivative or a
#                linear differential operator to be penalized.
#  Returns the penalty matrix.

#  Last modified 5 December 2001

if (!(inherits(basisfd, "basis.fd"))) stop(
    "First argument is not a basis.fd object.")

type <- getbasistype(basisfd)
if (type != "expon") stop ("Wrong basis type")

#  Find the highest order derivative in LFD

if (is.numeric(Lfd)) {
    if (length(Lfd) == 1) {
      	nderiv <- Lfd
      	if (nderiv != as.integer(nderiv)) {
        	stop("Order of derivative must be an integer")
      	}
      	if (nderiv < 0) {
        	stop("Order of derivative must be 0 or positive")
      	}
    } else {
      	stop("Order of derivative must be a single number")
    }
    if (nderiv < 0) stop ("Order of derivative cannot be negative")
} else if (inherits(Lfd, "fd")) {
   	derivcoef <- getcoef(Lfd)
   	if (length(dim(derivcoef))==3) derivcoef <- derivcoef[,,1]
   	nderiv <- dim(derivcoef)[2] - 1
   	if (nderiv < 0) {
   		stop("Order of derivative must be 0 or positive")
   	}
    nderiv <- ncol(derivcoef)
} else {
    stop("Second argument must be an integer or a functional data object")
}

#  Compute penalty matrix

if (is.numeric(Lfd)) {
    ratevec <- basisfd$params
    nrate   <- length(ratevec)
    penaltymatrix <- matrix(0,nrate,nrate)
    tl <- basisfd$rangeval[1]
    tu <- basisfd$rangeval[2]
    for (irate in 1:nrate) {
      	ratei <- ratevec[irate]
      	for (jrate in 1:irate) {
        	ratej <- ratevec[jrate]
        	ratesum <- ratei + ratej
        	if (ratesum != 0) {
          		penaltymatrix[irate,jrate] <- (ratei*ratej)^nderiv *
              	(exp(ratesum*tu) - exp(ratesum*tl)) / ratesum
        	} else {
          		if (nderiv == 0) penaltymatrix[irate,jrate] <- tu - tl
        	}
        	penaltymatrix[jrate,irate] <- penaltymatrix[irate,jrate]
      	}
	}
} else {
    penaltymatrix <- inprod(basisfd, basisfd, Lfd, Lfd)
}

return( penaltymatrix )
}

File: inst/Matlab/fdaM/ftpsite/temp/exponpen.m
function penaltymat = exponpen(basisobj, Lfdobj)
%EXPONPEN computes the exponential penalty matrix for penalty LFD.
%  Arguments:
%  BASISOBJ ... a basis.fd object of type 'expon'
%  LFDOBJ   ... a linear differential operator to be penalized. 
%               Default is 2.
%  Returns PENALTYMAT, the penalty matrix

%  Last modified:  9 September 2003

%  check BASISOBJ

if ~strcmp(class(basisobj), 'basis')
    error('First argument is not a basis object.')
end

%  check basis type

type = getbasistype(basisobj);
if ~strcmp(type, 'expon')
    error ('Basis type is not exponential.');
end

%  set up default linear differential operator

if nargin < 2, 
    Lfdobj = int2Lfd(2); 
end

%  check LFDOBJ

Lfdobj = int2Lfd(Lfdobj);

%  get highest order of derivative and check

nderiv = getnderiv(Lfdobj);
if nderiv < 0, error('NDERIV is negative.');  end

%  compute penalty matrix

if isinteger(Lfdobj)

    %  Compute penalty matrix exactly if the operator is D^NDERIV

    ratevec    = getbasispar(basisobj);
    nrate      = length(ratevec);
    penaltymat = zeros(nrate,nrate);
    rangeval   = getbasisrange(basisobj);
    tl = rangeval(1);
    tu = rangeval(2);
    for irate = 1:nrate
        ratei = ratevec(irate);
        for jrate = 1:irate
            ratej = ratevec(jrate);
            ratesum = ratei + ratej;
            if ratesum ~= 0
                penaltymat(irate,jrate) = (ratei*ratej)^nderiv * ...
                    (exp(ratesum*tu) - exp(ratesum*tl)) / ratesum;
            else
                if nderiv == 0
                    penaltymat(irate,jrate) = tu - tl;
                end
            end
            penaltymat(jrate,irate) = penaltymat(irate,jrate);
        end
    end
else
    
    %  LFDOBJ is not D^NDERIV, use approximate integration by calling
    %  function INPROD().
    
    penaltymat = inprod(basisobj, basisobj, Lfdobj, Lfdobj);
end

